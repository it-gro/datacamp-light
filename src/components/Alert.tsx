import * as React from "react";
import xss = require("xss");

import { IAlert } from "../redux/view";

import * as styles from "./Alert.module.scss";

export interface IAlertProps extends React.Props<Alert>, IAlert {
  className?: string;
  closable?: boolean;
}

export interface IAlertState {
  open: boolean;
}

export class Alert extends React.Component<IAlertProps, IAlertState> {
  public static defaultProps: Partial<IAlertProps> = {
    className: "",
    closable: false,
  };

  constructor(props: IAlertProps) {
    super(props);
    this.close = this.close.bind(this);
  }

  public componentWillMount() {
    this.setState({
      open: true,
    });
  }

  public close() {
    this.setState(prevState => ({
      ...prevState,
      open: false,
    }));
  }

  public componentWillReceiveProps() {
    this.setState(prevState => ({
      ...prevState,
      open: true,
    }));
  }

  public render() {
    let closeButton: JSX.Element = null;
    const { closable, className, type, content } = this.props;

    const sanitizedContent = xss(content, {
      stripIgnoreTag: true,
      stripIgnoreTagBody: true,
    });

    if (closable) {
      closeButton = (
        <button type="button" className={styles.close} onClick={this.close}>
          Ã—
        </button>
      );
    }

    return this.state.open ? (
      <div className={`${styles.alert} ${styles[type]} ${className}`}>
        {closeButton}
        <div dangerouslySetInnerHTML={{ __html: sanitizedContent }} />
      </div>
    ) : null;
  }
}

export default Alert;
