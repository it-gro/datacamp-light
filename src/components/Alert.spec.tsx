import * as React from "react";
import { shallow } from "enzyme";

import { Alert } from "./Alert";

describe("Alert", () => {
  it("should exist", () => {
    expect(Alert).toBeDefined();
  });

  it("should match snapshot", () => {
    const tree = shallow(<Alert />);
    expect(tree).toMatchSnapshot();
  });

  it("should prevent xss", async () => {
    const maliciousContent =
      '<img src="asd" onerror="window.dangerousFunction()" />' +
      '<script type="text/javascript">Hacky xss</script>';
    const component = shallow(
      <Alert type="danger" content={maliciousContent} />
    );
    const $ = component.render();
    const img = $.find("img");

    expect(img).toHaveLength(1);
    expect(img.attr("onerror")).toBeUndefined();
    expect($.find("script")).toHaveLength(0);
  });
});
