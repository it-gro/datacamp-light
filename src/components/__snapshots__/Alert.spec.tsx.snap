// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should match snapshot 1`] = `
<div
  className="alert undefined "
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
